### ISSUE TITLE
Concurrent Rendering

### ISSUE THREAD

**Issue by sunjay** [ProtoArt/spritec#116]
We currently take the simplest, most naive approach to executing our tasks:

https://github.com/ProtoArt/spritec/blob/40e562bb4b2c6a842322f8d1702de195aad1aa4d/src/bin/spritec/main.rs#L35-L45

We have a single loop that goes through each task one at a time, renders its images on the GPU, copies that data onto the CPU, and then writes out an image. Each individual render is very small (e.g. 32x32, 64x64, etc.), so that means that we spend an enormous amount of time on the CPU and very little time utilizing the GPU.

This issue is especially exacerbated for spritesheets because we render each frame on the GPU, then copy it into a buffer on the CPU where it then proceeds with laying it out into the final spritesheet image buffer. There is a separate issue open for this: #115 

One potential area to look at is to see if we can [use `PixelBuffer` instead](https://github.com/glium/glium/blob/221534849081e75625e1a14fe67a3effc97b0fb0/examples/screenshot-asynchronous.rs#L69-L70) of just [calling `read` and creating an image directly](https://github.com/ProtoArt/spritec/blob/40e562bb4b2c6a842322f8d1702de195aad1aa4d/src/renderer/thread_render_context.rs#L182). This may allow us to continue drawing while we download the previous drawing. 

Note that there would be some complications with this. For example, we want to be careful about using thread::spawn too much and potentially spawn a large number of threads, all waiting to download data from the GPU. It would be much better to have a single thread or some fixed N number of threads that wait on things. We would also need to time this and make sure we can actually download data and render at the same time in a cross platform way. Another complication with this comes from #114. If we're download the data asynchronously, we can't use that buffer again until we're done downloading. This essentially becomes a scheduling and resource sharing problem.

Another potential option to explore is to put more jobs together in the same image and run them at once on the GPU. As mentioned previously, our jobs are all relatively small in terms of resolution. We could potentially leverage the work in #115 to draw several jobs together (with clipping to keep things separate) and then split up the drawing on the CPU as we write out the images. To accomplish scaling in this setup, we could potentially create an image of the scaled size, render into unscaled sections of each tile, and then scale up in-place. No idea if the GPU would even be happy with that since it's probably a data race, but it's worth thinking about since it would allow us to do the entire render in a single texture.

We should also think about whether we can upload geometry data to the GPU concurrently while we render (cc #113). We should check if doing that is fast across all of our supported platforms.

**The most important thing to keep in mind with all of this is that we should collect and use timing data to drive our optimizations.** There is no point spending tons of time on a more complicated architecture if we don't end up making things faster.
